{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: biopython in /Users/work/Library/Python/3.10/lib/python/site-packages (1.83)\n",
      "Requirement already satisfied: scikit-learn in /Users/work/Library/Python/3.10/lib/python/site-packages (1.5.2)\n",
      "Requirement already satisfied: numpy in /Users/work/Library/Python/3.10/lib/python/site-packages (from biopython) (1.26.4)\n",
      "Requirement already satisfied: scipy>=1.6.0 in /Users/work/Library/Python/3.10/lib/python/site-packages (from scikit-learn) (1.14.1)\n",
      "Requirement already satisfied: joblib>=1.2.0 in /Users/work/Library/Python/3.10/lib/python/site-packages (from scikit-learn) (1.4.2)\n",
      "Requirement already satisfied: threadpoolctl>=3.1.0 in /Users/work/Library/Python/3.10/lib/python/site-packages (from scikit-learn) (3.5.0)\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m24.0\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m24.2\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpython3.10 -m pip install --upgrade pip\u001b[0m\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Running Task 1:\n",
      "Task 1 Results:\n",
      "\n",
      "R2 Score: 0.9506872648731928\n"
     ]
    }
   ],
   "source": [
    "# stuff only to run when not called via 'import' here\n",
    "%pip install biopython scikit-learn\n",
    "from sklearn.metrics import r2_score\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "import importlib\n",
    "import os\n",
    "import sys\n",
    "import time\n",
    "import alignment\n",
    "import copy\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio import pairwise2\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "\n",
    "print(\"Running Task 1:\")\n",
    "\n",
    "def CalculatePrimerFeatures(seq):\n",
    "    # modify this function to return a python list of feature values for a given sequence for Task 1\n",
    "    return [len(seq), seq.count('A'), seq.count('C'), seq.count('T'),seq.count('G'),seq.count('GC')]\n",
    "\n",
    "infile = open(\"training_primers.txt\", 'r')\n",
    "infile.readline() # don't load headers\n",
    "primers = []\n",
    "melting_points = []\n",
    "features = []\n",
    "st = time.time()\n",
    "for line in infile:\n",
    "    Line = line.split()\n",
    "    primers.append(Line[0])\n",
    "    melting_points.append(float(Line[1]))\n",
    "    # calculate features\n",
    "    features.append(CalculatePrimerFeatures(Line[0]))\n",
    "feat_time = (time.time()-st)/(len(features)/1000)\n",
    "# cross validation\n",
    "how_many_folds = 10 \n",
    "predictions = []\n",
    "truth = []\n",
    "my_len = len(features[-1])\n",
    "\n",
    "for fold in range(how_many_folds):\n",
    "    #print (\"Calculating Fold\",fold)\n",
    "    training_features = []\n",
    "    training_outcomes = []\n",
    "    testing_features = []\n",
    "    testing_outcomes = []\n",
    "    for c in range(len(melting_points)):\n",
    "        if c % how_many_folds == fold:\n",
    "            # put this one in testing data\n",
    "            testing_features.append(features[c])\n",
    "            testing_outcomes.append(melting_points[c])\n",
    "        else:\n",
    "            # put this one in training data\n",
    "            training_features.append(features[c])\n",
    "            training_outcomes.append(melting_points[c])\n",
    "    # train the model\n",
    "    \n",
    "    rf = RandomForestRegressor(n_estimators = 200)\n",
    "    rf.fit(training_features, training_outcomes)\n",
    "    fold_predictions = rf.predict(testing_features)\n",
    "    truth += testing_outcomes\n",
    "    predictions += list(fold_predictions)\n",
    "  \n",
    "        \n",
    "#truth = np.array(truth)\n",
    "#predictions = np.array(predictions)        \n",
    "print(\"Task 1 Results:\\n\")\n",
    "print(\"R2 Score:\", r2_score(truth, predictions))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "\"\"\"\n",
    "Created on Thu Oct  4 17:32:48 2018\n",
    "\n",
    "@author: jdkan\n",
    "\"\"\"\n",
    "import time\n",
    "import alignment\n",
    "import copy\n",
    "\n",
    "# Your task is to *accurately* predict the primer melting points using machine \n",
    "# learning based on the sequence of the primer.\n",
    "\n",
    "# Load the primers and their melting points.\n",
    "\n",
    "\n",
    "def CalculatePrimerFeatures(seq):\n",
    "    # modify this function to return a python list of feature values for a given sequence for Task 1\n",
    "    return [len(seq), seq.count('A'), seq.count('C'), seq.count('T'),seq.count('G'),seq.count('GC')]\n",
    "\n",
    "\n",
    "def GetMeltingPoint(primer:str,melting_point_rf):\n",
    "    return melting_point_rf.predict([CalculatePrimerFeatures(primer)])\n",
    "    \n",
    "# template sequence is 5' -> 3'\n",
    "# primer is 3' -> 5'\n",
    "\n",
    "def get_smallest_index_match_on_template_sequence(template_sequence:str, primer:str):\n",
    "    scoring = alignment.ScoreParam(10,-5,-7)\n",
    "    # binding if at least 80% max score\n",
    "    SCORE_THRESHOLD = .8 * (scoring.match*len(primer))\n",
    "    best, optloc, matrix = alignment.local_align(template_sequence, primer, score=scoring)\n",
    "    if best < SCORE_THRESHOLD:\n",
    "        return 9999999\n",
    "\n",
    "    return optloc[0]-1 # the more straight-forward approach\n",
    "\n",
    "    for i in range(len(template_sequence)+1):\n",
    "        for j in range(len(primer),-1,-1):\n",
    "            if(matrix[i][j]>=SCORE_THRESHOLD):\n",
    "                print(i,j,matrix[i][j])\n",
    "                return max(0,i-1) # includes primer\n",
    "    return 9999999 # MAX_INT\n",
    "            \n",
    "\n",
    "def reverse_complement(primer:str):\n",
    "    complement = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C'}\n",
    "    comp_match = ''.join(complement[base] for base in reversed(primer))\n",
    "    return comp_match\n",
    "\n",
    "def complement(primer:str):\n",
    "    complement = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C'}\n",
    "    comp_match = ''.join(complement[base] for base in primer)\n",
    "    return comp_match\n",
    "\n",
    "def PredictPCRProduct(primer1, primer2, template_sequence, melting_point_rf):\n",
    "    \"\"\"\n",
    "    Input:\n",
    "        primer1 = a primer sequence in 5' to 3' order\n",
    "        primer2 = a primer sequence in 5' to 3' order\n",
    "        template_sequence = sequence from which we are trying to generate \n",
    "        copies using PCR in 5' to 3' order.  Assume this is double stranded, \n",
    "        but we are only including the top strand in the argument.\n",
    "        melting_point_rf = random forest learned from task1 to predict primer\n",
    "        melting points.\n",
    "    Output:\n",
    "        return sequence of successful PCR amplication reaction or None (if \n",
    "        there is no successful reaction)\n",
    "        \n",
    "    \"\"\"\n",
    "    #Fail any primers with length out of range 18-30 bp\n",
    "    if(not ((18 <= len(primer1) <= 30) and (18 <= len(primer2) <= 30))):\n",
    "        print(\"Lengths not right!\",len(primer1),len(primer2))\n",
    "        return None\n",
    "        \n",
    "    #Fail any primers with conflicting melting points \n",
    "    melting_point_1, melting_point_2 = GetMeltingPoint(primer1,melting_point_rf), GetMeltingPoint(primer2,melting_point_rf)\n",
    "    if abs(melting_point_1-60)>2 or abs(melting_point_2-60)>2:\n",
    "        print(\"temps not right\",melting_point_1,melting_point_2)\n",
    "        return None\n",
    "        \n",
    "    #Check for binding on both primers\n",
    "    top_strand_match = min(\n",
    "        get_smallest_index_match_on_template_sequence(template_sequence = template_sequence, primer = reverse_complement(primer1)),\n",
    "        get_smallest_index_match_on_template_sequence(template_sequence = template_sequence, primer = reverse_complement(primer2)))\n",
    "    \n",
    "    bottom_strand_match = len(template_sequence)-1-(\n",
    "        min(get_smallest_index_match_on_template_sequence(template_sequence = reverse_complement(template_sequence), primer = reverse_complement(primer1)),\n",
    "            get_smallest_index_match_on_template_sequence(template_sequence = reverse_complement(template_sequence), primer = reverse_complement(primer2)))) # reverse the primers\n",
    "\n",
    "    if top_strand_match-bottom_strand_match<=0 or bottom_strand_match<0 or top_strand_match > 99999:\n",
    "        return None # no overlap\n",
    "    \n",
    "    \n",
    "\n",
    "    return template_sequence[bottom_strand_match:top_strand_match+1]\n",
    "\n",
    "\n",
    "def LoadFastA(path):\n",
    "    infile = open(path, 'r')\n",
    "    seq = \"\"\n",
    "    infile.readline()\n",
    "    for line in infile:\n",
    "        seq += line[:-1]\n",
    "    return seq\n",
    "    \n",
    "if __name__ == \"__main__\":\n",
    "\n",
    "\n",
    "   \n",
    "   \"\"\"\n",
    "   Task 2:\n",
    "   Design a function to predict whether a product will be made in a PCR reaction.\n",
    "   Your function should take as input the template DNA and the two primers and \n",
    "   return the product or 'None'.\n",
    "\n",
    "   This requires a local alignment function which is provided for you or you \n",
    "   can use another implementation.\n",
    "    \n",
    "   There are test cases in PCR_product_test_cases.txt.\n",
    "   \n",
    "   \"\"\"\n",
    "   task2_randomforest = RandomForestRegressor(n_estimators = 200)\n",
    "   task2_randomforest.fit(features, melting_points)\n",
    "   \n",
    "#    print(PredictPCRProduct(\"ACTG\", \"ACTG\", \"ACTCAGCGACTGC\", task2_randomforest))\n",
    "   assert PredictPCRProduct(\"TGGTGGGATGTCTTTCAACAGG\", \"AACTACGGAGAACTACAGCAACCT\",\"ACGTCAGCGAGCGCTACGACGTGGTGGGATGTCTTTCAACAGGACGGACTGACGCGACGACTGACTGTAGGCTAGGTTGCTGTAGTTCTCCGTAGTTAGCTACGACGCATGCAGCTGCA\", task2_randomforest) == \"TGGTGGGATGTCTTTCAACAGGACGGACTGACGCGACGACTGACTGTAGGCTAGGTTGCTGTAGTTCTCCGTAGTT\"\n",
    "   \n",
    "   \n",
    "   \"\"\"\n",
    "   Task 3:\n",
    "   Design primers for a PCR reaction to distinguish between the three (strands)\n",
    "   types of DNA.  \n",
    "   \n",
    "   -Your primers should be between 18 and 35 bases long.  \n",
    "   -They should have at least 80% match to the DNA strand.\n",
    "   -Predicted melting points of any primers to be run in the same reaction\n",
    "   should be between 58.0 and 62.0 C.\n",
    "   -Products are distinguishable in length if their difference in length is >40\n",
    "   bases\n",
    "   -Your products should not be longer than 1000 bases.\n",
    "   \n",
    "   We are making predictions about the functionality of sets of primers.  We \n",
    "   will synthesize your group's primers and test them in the lab later.\n",
    "     \n",
    "   \"\"\"\n",
    "\n",
    "def generatePrimers(sequences):\n",
    "    #gattacas a bad movie\n",
    "    primers = [(\"GATTACA\",\"ACTACAT\")]\n",
    "\n",
    "\n",
    "    return primers\n",
    "   \n",
    "   \n",
    "\n",
    "# TGCAGTCGCTCGCGATGCTGCACCACCCTACAGAAAGTTGTCCTGCCTGACTGCGCTGCTGACTGACATCCGATCCAACGACATCAAGAGGCATCAATCGATGCTGCGTACGTCGACGT AGGTTGCTGTAGTTCTCCGTAGTT\n",
    "\n",
    "# TGCAGTCGCTCGCGATGCTGCACCACCCTACAGAAAGTTGTCCTGCCTGACTGCGCTGCTGACTGACATCCGATCCAACGACATCAAGAGGCATCAATCGATGCTGCGTACGTCGACGT GGACAACTTTCTGTAGGGTGGT\n",
    "# ACGTCAGCGAGCGCTACGACGTGGTGGGATGTCTTTCAACAGGACGGACTGACGCGACGACTGACTGTAGGCTAGGTTGCTGTAGTTCTCCGTAGTTAGCTACGACGCATGCAGCTGCA GGACAACTTTCTGTAGGGTGGT\n",
    "\n",
    "# ACGTCAGCGAGCGCTACGACGTGGTGGGATGTCTTTCAACAGGACGGACTGACGCGACGACTGACTGTAGGCTAGGTTGCTGTAGTTCTCCGTAGTTAGCTACGACGCATGCAGCTGCA \n",
    "#                                                    AGGTTGCTGTAGTTCTCCGTAGTT\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
